pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+	/// `step()` will be made. Any value less than `Perbill::one()` will result in another future
+	/// call to `step()`. Indeed, values < 1 are arbitrary, but the intent is to indicate progress
+	/// (so they should at least be monotonically increasing).
+	fn step(&self, previous_progress: Perbill, available_weight: Weight) -> (Perbill, Weight);
+}
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_std::prelude::*;
+
+	/// Pallet for migrations
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The list of migrations that will be performed
+		type MigrationsList: Get<Vec<Box<dyn Migration>>>;
+		/// Whether or not multi-block migrations are supported
+		type MultiBlockMigrationsSupported: Get<bool>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RuntimeUpgradeStarted(),
+		RuntimeUpgradeStepped(Weight),
+		RuntimeUpgradeCompleted(),
+		MigrationStarted(Vec<u8>),
+		MigrationStepped(Vec<u8>, Perbill, Weight),
+		MigrationCompleted(Vec<u8>),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// on_runtime_upgrade is expected to be called exactly once after a runtime upgrade.
+		/// We use this as a chance to flag that we are now in upgrade-mode and begin our
+		/// migrations.
+		///
+		/// In the event that a migration is expected to take more than one block, ongoing migration
+		/// work could continue from block-to-block in this pallet's on_initialize function.
+		fn on_runtime_upgrade() -> Weight {
+			log::warn!("Performing on_runtime_upgrade");
+
+			let mut weight: Weight = 0u64.into();
+			// TODO: derive a suitable value here, which is probably something < max_block
+			let available_weight: Weight = T::BlockWeights::get().max_block;
+
+			// start by flagging that we are not fully upgraded
+			// TODO: case where this is already false (e.g. we started an upgrade while one was
+			// already in progress)
+			//		notably, we might have started an individual migration and the list of
+			//		migrations might change on our next on_runtime_upgrade()
+			<FullyUpgraded<T>>::put(false);
+			weight += T::DbWeight::get().writes(1);
+			Self::deposit_event(Event::RuntimeUpgradeStarted());
+
+			weight += process_runtime_upgrades::<T>(available_weight - weight);
+
+			// at least emit a warning if we aren't going to end up finishing our migrations...
+			if !T::MultiBlockMigrationsSupported::get() {
+				if !<FullyUpgraded<T>>::get() {
+					log::warn!(
+						"migrations weren't completed in on_runtime_upgrade(), but we're not
+					configured for multi-block migrations; state is potentially inconsistent!"
+					);
```

Comment:
```suggestion
					log::error!(
						"migrations weren't completed in on_runtime_upgrade(), but we're not
					configured for multi-block migrations; state is potentially inconsistent!"
					);
```
should probably be error